PIPS/pip-1.md

```diff
@@ -0,0 +1,147 @@
+    bytes nativeAssetHash;
+}
+
+mapping(address => DelegatedAsset) delegatedAssets;
+mapping(bytes32 => uint256) balances;
+
+function delegateAsset(uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash, uint256 totalSupply) public {
+    address assetHash = _msgSender();
+
+    require(nativeChainId != 0);
+    require(delegatedAssets[assetHash].nativeChainId == 0);
+
+    // `hash` is any supported hashing function of the respective blockchain
+    bytes32 key = hash(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+
+    require(balances[key] == 0);
+    balances[key] = totalSupply;
+
+    delegatedAssets[_msgSender()] = DelegatedAsset(nativeChainId, nativeLockProxy, nativeAssetHash);
+
+    // call registerAsset here for convenience
+    // registerAsset is explained in a later section
+    registerAsset(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+}
+```
+
+This mapping will be used in the `unlock` function to ensure that delegated tokens will be unlocked only if the specified token is locked.
+
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed, to replace them we can add a `registerAsset` function:
+
+```
+mapping(bytes32 => bool) registry;
+
+function registerAsset(address fromAssetHash, uint64 toChainId, bytes memory targetProxyHash, bytes memory toAssetHash) {
+    key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
+    registry[key] = true;
+}
+```
+
+This function can be permissionless as its main purpose is to help prevent careless mistakes from the user when calling the `lock` function.
+
+The `lock` function can be modified to require the following parameters:
+- address fromAssetHash
+- uint64 toChainId
+- bytes memory targetProxyHash
+- bytes memory toAssetHash
+- bytes memory toAddress
+- uint256 amount
+
+The LockProxy contract should maintain a `balances` mapping of (bytes32 => uint256).
+This `balances` mapping should be updated within the `lock` function:
+```
+// help prevent user mistakes by ensuring delegated assets only get transferred
+// to the native chainId, LockProxy and assetHash that was originally specified
+DelegatedAsset dAsset = delegatedAssets[fromAssetHash];
+if (dAsset.nativeChainId != 0) {
+    require(
+        dAsset.nativeChainId == toChainId &&
+        dAsset.nativeLockProxy == targetProxyHash &&
+        dAsset.nativeAssetHash == toAssetHash
+    );
+}
+
+bytes32 key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
+require(registry[key] == true);
+
+balances[key] += amount;
+
+TxArgs memory txArgs = TxArgs({
+    fromContractAddr: address(this),
+    fromAssetHash: fromAssetHash,
+    toAssetHash: toAssetHash,
+    toAddress: toAddress,
+    amount: amount
+});
+bytes memory txData = _serializeTxArgs(txArgs);
+
+emit LockEvent(txData);
```

Comment:
i think this notify info can update to:
```
LockEvent(fromContractAddr, fromAssetHash, toChainId, toAssetHash, toAddress, amount)
```

Reply:
it will more friendly for front-end develop.

Reply:
that's true, I'll update it